src/Java.Interop.Tools.Maven/Repositories/CachedMavenRepository.cs

```diff
@@ -0,0 +1,94 @@
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Java.Interop.Tools.Maven.Models;
+
+namespace Java.Interop.Tools.Maven.Repositories;
+
+/// <summary>
+/// Wraps an <see cref="IMavenRepository"/> and caches files in a local directory.
+/// </summary>
+public class CachedMavenRepository : IMavenRepository
+{
+	public string CacheDirectory { get; }
+
+	public string Name => repository.Name;
+
+	readonly IMavenRepository repository;
+
+	public CachedMavenRepository (string directory, IMavenRepository repository)
+	{
+		CacheDirectory = directory;
+		this.repository = repository;
+	}
+
+	public bool TryGetFile (Artifact artifact, string filename, [NotNullWhen (true)] out Stream? stream)
+	{
+		stream = null;
+
+		if (TryGetFilePath (artifact, filename, out var path)) {
+			stream = File.OpenRead (path);
+			return true;
+		}
+
+		return false;
+	}
+
+	public bool TryGetFilePath (Artifact artifact, string filename, [NotNullWhen (true)] out string? path)
+	{
+		path = null;
+
+		var directory = GetArtifactDirectory (artifact);
+		var file = Path.Combine (directory, filename);
+
+		if (File.Exists (file)) {
+			path = file;
+			return true;
+		}
+
+		Directory.CreateDirectory (directory);
+
+		if (repository.TryGetFile (artifact, filename, out var repo_stream)) {
+			using (var sw = File.Create (file))
+				repo_stream!.CopyTo (sw);
+
+			using (repo_stream) { }
```

Comment:
This feels weird.

I think this would feel better:

```csharp
using (var sw = File.Create (file))
using (repo_stream) {
    repo_stream.CopyTo (sw);
}
```

Reply:
Indeed that is better.